### ISSUE TITLE
trainer resume from checkpoint，the learning rate is not the same as retraining,learning rate is discontinuous

### ISSUE THREAD

**Issue by LBJ6666** [huggingface/transformers#34053]
### System Info

- Platform: Windows-10
- transformers version: 4.43.4
- Python version: 3.10.11
- PyTorch version (GPU?): 2.3.1+cu121

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

the Trainer does not set a warmup and the lr_scheduler is set to linear, and the training is continued from an interruption to complete all steps, the learning rates will be different from those when training all steps from the beginning. Here are the specific learning rates:

Learning rates for training from the beginning for each step:
- Step 1: "learning_rate": 1e-05,
- Step 2: "learning_rate": 1e-05,
- Step 3: "learning_rate": 9e-06,
- Step 4: "learning_rate": 8.000000000000001e-06,
- Step 5: "learning_rate": 7e-06,
- Step 6: "learning_rate": 6e-06,
- Step 7: "learning_rate": 5e-06,
- Step 8: "learning_rate": 4.000000000000001e-06,
- Step 9: "learning_rate": 3e-06,
- Step 10: "learning_rate": 2.0000.

If training is continued from a checkpoint at step 5, the learning rates for each step are:
- Step 6: "learning_rate": 7e-06,
- Step 7: "learning_rate": 7e-06,
- Step 8: "learning_rate": 6e-06,
- Step 9: "learning_rate": 5e-06,
- Step 10: "learning_rate": 4.000000000000001e-06.

Why are the learning rates for step 6 and step 7 different when training continues from a checkpoint compared to training from the start?

Reproduction steps:
1. Train from the beginning for 10 steps, save a checkpoint for each step, and record the learning rate in each step.
2. Delete the checkpoints for steps 6 through 7 in the folder.
3. Then use `trainer.train(resume_from_checkpoint=True)` to continue training from step 5, and after training is completed, record the learning rate in the new checkpoint.

### Expected behavior

Please explain why the learning rate is not continuous as it is when training from the beginning, for example:
Step 6: "learning_rate": 6e-06,
Step 7: "learning_rate": 5e-06.
........

**Comment by LysandreJik** [comment]
cc @muellerzr @SunMarc

**Comment by LysandreJik** [comment]
I'm adding a "Good First Issue" tag; the team is under tight bandwidth at the moment so any PR to help solve this is welcome.

**Comment by SunMarc** [comment]
Hey @LBJ6666, could you please share a minimal reproducer, so that we can quickly find the issue ? Thanks !

**Comment by LBJ6666** [comment]
@SunMarc ,Thank for response,I wrote a simple example.
```
import os
from datasets import load_dataset
from transformers import AutoTokenizer, AutoModelForCausalLM, Trainer, TrainingArguments
import numpy as np

def tokenize_function(examples):
    model_inputs = tokenizer(examples["prompt"], padding="max_length", truncation=True)
    model_inputs["labels"] = np.array(model_inputs['input_ids']).astype(np.int64)
    return model_inputs

dataset = load_dataset("fka/awesome-chatgpt-prompts")

model = AutoModelForCausalLM.from_pretrained("openai-community/gpt2")
tokenizer = AutoTokenizer.from_pretrained("openai-community/gpt2")
tokenizer.pad_token = tokenizer.eos_token
tokenized_datasets = dataset.map(tokenize_function, batched=True)

output_dir = "./output"
training_args = TrainingArguments(
    output_dir=output_dir,
    fp16=True,
    max_steps=20,
    per_device_train_batch_size=4,
    learning_rate=1e-5,
    lr_scheduler_type="cosine",
    logging_dir=os.path.join(output_dir, "log"),
    logging_strategy="steps",
    logging_steps=1,
    report_to="tensorboard",
    save_strategy='steps',
    save_steps=1,
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets['train'],
)

trainer.train(
    resume_from_checkpoint=False
)
trainer.save_state()

```
Run the training and record the learning rates for 20 steps as saved in `trainer_state.json` within `checkpoint-20`.  
then, delete checkpoints 16-20 to simulate an interrupted training process.  
set `resume_from_checkpoint=True` to resume training from step 16, and compare the learning rates in the final `trainer_state.json` after completing the training.

or
see tensorboard log

**Comment by Knight7561** [comment]
Verify my steps of Reproduction:
For the sample example above:
1. Case A: Ran trainer for 20 epochs and saved JSON log.
2.  Case B: Deleted checkpoints -15 and 16, and reran the trainer from checkpoint-14.
However, I found the LR rate was the same in both cases. Please review the screenshot and let me know if I am missing any step in reproduction.

![Screenshot 2024-11-21 at 12 09 35](https://github.com/user-attachments/assets/4107dac3-68bc-46ea-9290-b5674ad96f8d)

**Comment by SunMarc** [comment]
Thanks for testing this @Knight7561 ! Could you check @LBJ6666 ?

**Comment by SunMarc** [comment]
Is there anything that could have been added to make this easier to spot ? Maybe some logs ?

**Comment by hsilva664** [comment]
Hello, I tried reproducing the issue with fp16 and the undesired behaviour was present. Code is as posted before, except for a new part I've added for setting the seed equal to zero. I tried first running the 20 steps while saving results, then deleting steps 12 onwards, switching the flag `resume_from_checkpoint` to `True` and running again.

As reported by @LBJ6666 , when there is gradient overflow, the optimizer does not run and, as a consequence, the LR scheduler does not get updated. However, when reproducing, the issue was a little worse than that in my case. When running the 20 steps without interruption, gradient overflow did not occur on my machine. However, it did occur when I loaded checkpoint 11.

After some inspection, I noticed that there was an object from a class called `GradScaler` which was an attribute of another object from the `Accelerator` class. When running from scratch, the scaler multiplies the loss from the step 11 model by `256` before backpropagating. However, when loading from file, the scaler actually multiplies it by `65536` instead, which causes the overflow. To solve it, I tried adding a function to save and load the scaler as well, similarly to how the scheduler and optimizer are saved and loaded. It seems to work more consistently.

Still, there are some nuances to saving and reloading that I'm not sure how to handle, such as when using Deepspeed, Sagemaker or XLA. I just copied what was done for the scheduler for now and left the code as a draft so others can look at and say how to best change it (the structure is a little unnecessarily duplicated, but this was, again, just a draft to see if it would work). Also, when loading the `GradientScaler`, the performance between loaded and from-scratch models gets very similar, but it does nor become exactly equal, as I think it should. I've opened the pull request at https://github.com/huggingface/transformers/pull/34932#issue-2692506485

**Comment by SunMarc** [comment]
> After some inspection, I noticed that there was an object from a class called GradScaler which was an attribute of another object from the Accelerator class. When running from scratch, the scaler multiplies the loss from the step 11 model by 256 before backpropagating. However, when loading from file, the scaler actually multiplies it by 65536 instead, which causes the overflow. To solve it, I tried adding a function to save and load the scaler as well, similarly to how the scheduler and optimizer are saved and loaded. It seems to work more consistently.

Thanks for the nice investigation. This looks indeed like a bug. cc @muellerzr 

Could you check @Knight7561 and @LBJ6666 that this is potentially the issue ?

**Comment by Knight7561** [comment]
Sure @SunMarc , let me debug this and try to reproduce the above case scenario to see what's going on.

**Comment by Knight7561** [comment]
Thanks for the results @hsilva664. So, if I understand corrected, once you ran uninterrupted(got no `inf` except the starting epochs) and then you deleted the checkpoint-11 and then you rerank with `resume_from_checkpoint to True`, which lead to `inf` again (which is unexpected behaviour after cp11), right? and then when you fixed the scaler issue in your PR, thats when you get proper results as in Screenshot 3?

Am I understanding correctly? @hsilva664

**Comment by hsilva664** [comment]
I did not delete checkpoint 11, only checkpoints 12 onwards. Other than this, your explanation is correct.

**Comment by LBJ6666** [comment]
@hsilva664 Thank you for the further testing.

I have successfully tested the code on the PR #34932  , and the results show that the logs for first running the 20 steps of training and the logs for resuming training from checkpoint 11 (train steps 12-20) are identical, including the loss, gradient, and learning rate.

Regarding result reproducibility, the slight differences observed between two runs on the same machine are due to non-determinism. To ensure consistent results, the following configuration changes can reproduce the results:
```
import os
os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":16:8"
import numpy as np
import random
import torch

def set_seed(seed):
    random.seed(seed)
    np.random.seed(seed)
    torch.manual_seed(seed)
    if torch.cuda.is_available():
        torch.cuda.manual_seed_all(seed)
    torch.use_deterministic_algorithms(True)
set_seed(0)
```

**Comment by Knight7561** [comment]
So, @LBJ6666, Does the PR solves your Issue?

**Comment by hsilva664** [comment]
Hello, by previous discussions, it seems so. Still, the PR is not ready to be merged. As per request, I tried adding some tests that would detect the save/reloading errors better than the tests that were already there. However, these new tests required setting deterministic behaviour differently, which can be done via flags and environment variables. My new tests generate consistent results when run in isolation or when run together with the tests from class `TrainerIntegrationTest`, but they do not pass when all tests from the containing file are run together.

I assume this might be because another test, from another class but same file, running on some parallel worker somehow changes the global state in a way that the deterministic behaviour I had set stopped working. Since I'm not sure how to proceed, I left it here until somebody takes a look. More details are in the PR discussion.

**Comment by LBJ6666** [comment]
@Knight7561 ，Yes, after testing my code, it resolved my issue. I look forward to more comprehensive testing.
